Add user list component + snap to user functionality

oliverbenns opened the pull request:
- Creates a horizontal user avatar list in the top right of the window that displays the current users, with the first 2 letters of the username if provided, else a `?` appears. Exact labelling/naming tbd. The design could probably be improved, looking for feedback or ideas.
- Adds functionality, when clicking a collaborator's avatar, snaps the user to their current cursor position.

This somewhat solves the problem of https://github.com/excalidraw/excalidraw/issues/1721 . The solution provided here is a first pass by snapping the user's window to center on the target collaborators cursor. No following occurs as this will require adding collaborator scroll position to state and sending this over sockets which is a larger chunk of work.

There shouldn't be too many surprises with the code here, with the notable exception of this being the first time where we are rendering actions dynamically, and therefore need to pass a unique key in order to pick props off the state, which combats the need to get into messing generic props which is probably overkill at this stage.

How it looks:
<img width="1247" alt="Screenshot 2020-06-10 at 23 15 02" src="https://user-images.githubusercontent.com/1992355/84324252-44f7d180-ab70-11ea-96ce-253ce656c14a.png">

Interaction clicking between collaborators:


![Large GIF (1632x1038)](https://user-images.githubusercontent.com/1992355/84324319-6c4e9e80-ab70-11ea-96ed-85fbef4b89b9.gif)

dwelle commented on src/clients.ts:
```diff
@@ -0,0 +1,19 @@
+
+// @TODO: This is terrible for i18n. Also we probably want to get initials if words len > 1?
+export const getClientShortName = (username: string) => {
+  return username.substring(0, 2);
```
I'd use initials if user supplies first & last name, and only then fallback to first two chars. Also, let's uppercase the letters in either scenario.

oliverbenns replied:
Yeah I thought this too but for international names are we ok with doing the same? Does that make sense in all languages? What if a user inputs a middle name (so 3 in total)?

As I don't know these answers I thought a substring might be safer. Uppercase is a good call though. Thoughts?

dwelle replied:
I think it's still safe to go with initials first. If someone uses 3 names (unlikely), we can take first and last. I don't see much harm even if we get it wrong in some edge case, and it'll generally improve the UX.

dwelle commented on src/components/Avatar.tsx:
```diff
@@ -0,0 +1,25 @@
+  color: string;
+};
+
+export const Avatar = ({
```
Let's take `title` prop and display `Unnamed collaborator` on hover if user doesn't supply username.

oliverbenns replied:
Yeah hover sounds good - showing either full username or unnamed.

oliverbenns replied:
Added tooltip component which wraps around the avatars. Title is a weird name for the prop - changed children to enforce string-only but happy to make the prop 'initials' or something.

dwelle commented:
Nice work!

ad1992 commented on src/components/FixedUserList.css:
```diff
@@ -0,0 +1,13 @@
+.FixedUserList {
+  --margin: 0.25rem;
+  position: fixed;
```
can you use position absolute instead so that its positioned with respect to its parent instead of viewport?

ad1992 commented on src/components/FixedUserList.tsx:
```diff
@@ -0,0 +1,12 @@
+  className?: string;
+};
+
+export const FixedUserList = ({ children, className }: FixedUserListProps) => (
```
can we rename it to just `UserList` ? Not sure if Fixed makes sense!

lipis commented:
How does it scale with more participants or smaller screen?

oliverbenns commented:
Few things updated:
- Fixed initials
- Added tooltip component. 
- Removed fixed positioning and put in app menu.
- Added responsive styles so the avatar stack when there are a ton of collaborators (this creates a long vertical list on small, portrait tablets but this is an extreme edge case with _loads_ of users) 
- Added collaborators to mobile menu. Mobile menu closes on touch of a collaborator.

**Tooltips:**
<img width="201" alt="Screenshot 2020-06-11 at 23 28 28" src="https://user-images.githubusercontent.com/1992355/84445739-0f201f00-ac3c-11ea-916e-c0870cf695e5.png">
<img width="124" alt="Screenshot 2020-06-11 at 23 31 46" src="https://user-images.githubusercontent.com/1992355/84445755-10514c00-ac3c-11ea-95c0-5407a2003119.png">

**Small screens:**
<img width="816" alt="Screenshot 2020-06-11 at 23 29 52" src="https://user-images.githubusercontent.com/1992355/84445850-23641c00-ac3c-11ea-94d6-6b26df8c3c9b.png">

**Mobile menu:**

<img width="388" alt="Screenshot 2020-06-11 at 23 30 50" src="https://user-images.githubusercontent.com/1992355/84445938-64f4c700-ac3c-11ea-8cea-e49062668a99.png">

oliverbenns commented:
I believe I have addressed the feedback here - what are the next steps in getting this merged?

dwelle commented:
Sorry, I completely blocked this out. Looks great, thanks a lot!
